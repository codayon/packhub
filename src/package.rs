use std::sync::{Arc, Mutex};

use anyhow::{bail, Result};
use chrono::{DateTime, Utc};

use crate::utils::{Dist, Type};

struct InnerPackage {
    tipe: Type,
    dist: Option<Dist>,
    url: String,
    ver: String,
    data: Mutex<Data>,
    created: DateTime<Utc>,
    updated: DateTime<Utc>,
}

#[derive(Clone, PartialEq)]
pub struct Package {
    inner: Arc<InnerPackage>,
}

/// Data type for package data and metadata.
///
/// It is used to differentiate between package data and metadata.
///
/// `Data::Package` is used for package data. It is the actual package
/// file which needs to be processed to extract the metadata.
///
/// `Data::Metadata` is used for package metadata.
///
/// `Data::None` is used when no data is available.
#[derive(Clone, PartialEq)]
pub enum Data {
    Package(Vec<u8>),
    Metadata(String),
    None,
}

impl std::fmt::Debug for Package {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.file_name())
    }
}

impl PartialEq for InnerPackage {
    fn eq(&self, other: &Self) -> bool {
        self.tipe == other.tipe
            && self.dist == other.dist
            && self.url == other.url
            && self.ver == other.ver
            && *self.data.lock().unwrap() == *other.data.lock().unwrap()
            && self.created == other.created
            && self.updated == other.updated
    }
}

impl Package {
    pub fn detect_package(
        name: &str,
        ver: String,
        url: String,
        created: DateTime<Utc>,
        updated: DateTime<Utc>,
    ) -> Result<Package> {
        // Split the extension first.
        // If we don't recognize it, then return error.
        let Some((tipe, splitted)) = split_extention(name) else {
            bail!("Unknown package type: {}", name);
        };

        let mut dist: Option<Dist> = None;
        let sections: Vec<&str> = splitted.split(['-', '_']).collect();

        for section in sections {
            match section {
                dst if dst.contains("ubuntu") => dist = Some(Dist::Ubuntu(parse_version(dst))),
                dst if dst.contains("debian") => dist = Some(Dist::Debian(parse_version(dst))),
                dst if dst.contains("fedora") => dist = Some(Dist::Fedora(parse_version(dst))),
                _ => (),
            }
        }

        let inner = InnerPackage {
            tipe,
            dist,
            url,
            ver,
            data: Mutex::new(Data::None),
            created,
            updated,
        };

        Ok(Self {
            inner: Arc::new(inner),
        })
    }

    pub fn ty(&self) -> &Type {
        &self.inner.tipe
    }

    /// Return the distribution for which it was packaged
    pub fn distribution(&self) -> &Option<Dist> {
        &self.inner.dist
    }

    /// Version of the package
    pub fn version(&self) -> &str {
        &self.inner.ver
    }

    pub fn download_url(&self) -> &str {
        &self.inner.url
    }

    pub fn file_name(&self) -> &str {
        &self.inner.url.split('/').last().unwrap()
    }

    /// Download package data
    ///
    /// It is required to call this function before calling the `data()` function.
    pub async fn download(&self) -> Result<()> {
        let data = reqwest::get(self.download_url()).await?.bytes().await?;
        *self.inner.data.lock().unwrap() = Data::Package(data.to_vec());
        Ok(())
    }

    /// Return the data of the package.
    ///
    /// It is required to call the `download()` or `set_metadata()` function before calling this.
    /// Otherwise, `None` is returned.
    pub fn data(&self) -> Data {
        self.inner.data.lock().unwrap().clone()
    }

    #[cfg(test)]
    /// Set the package data.
    ///
    /// It's for testing purpose.
    pub fn set_package_data(&self, data: Vec<u8>) {
        *self.inner.data.lock().unwrap() = Data::Package(data);
    }

    pub fn creation_date(&self) -> &DateTime<Utc> {
        &self.inner.created
    }

    pub fn updated_date(&self) -> &DateTime<Utc> {
        &self.inner.updated
    }

    /// Set the package metadata.
    pub fn set_metadata(&self, metadata: String) {
        *self.inner.data.lock().unwrap() = Data::Metadata(metadata);
    }

    /// Check if metadata is available.
    pub fn is_metadata_available(&self) -> bool {
        matches!(*self.inner.data.lock().unwrap(), Data::Metadata(_))
    }
}

/// Parses the version from the distribution identifier `dist`.
///
/// For instance, for a distribution identifier `ubuntu22.10` it will
/// parse the version as `22.10`.
fn parse_version(dist: &str) -> Option<String> {
    split_at_numeric(dist).map(|s| s.to_owned())
}

/// Splits the string `s` at the first occurence of a numeric digit.
///
/// It is used to extract version number from strings, such as for "ubuntu24.10" it would
/// return "24.10".
fn split_at_numeric(s: &str) -> Option<&str> {
    for (curr, (index, next)) in s.chars().zip(s.char_indices().skip(1)) {
        if curr.is_ascii_alphabetic() && next.is_ascii_digit() {
            return Some(&s[index..]);
        }
    }

    None
}

fn split_extention(s: &str) -> Option<(Type, &str)> {
    let mut str = String::with_capacity(3);
    let mut index = 0;

    for (idx, ch) in s.char_indices().rev() {
        if ch == '.' {
            index = idx;
            break;
        } else {
            str.push(ch);
        }
    }

    if index == 0 {
        return None;
    }

    let splitted = &s[0..index];

    // `str` is in reverse order, so we try to match it reversely.
    let tipe = match str.as_str() {
        "bed" => Type::Deb,
        "mpr" => Type::Rpm,
        _ => return None,
    };

    Some((tipe, splitted))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_package() {
        let pack = Package::detect_package(
            "OpenBangla-Keyboard_2.0.0-ubuntu22.04.deb",
            "2.0.0".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
            DateTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(pack.version(), "2.0.0");
        assert_eq!(
            *pack.distribution(),
            Some(Dist::Ubuntu(Some("22.04".to_owned())))
        );
        assert_eq!(*pack.ty(), Type::Deb);

        let pack = Package::detect_package(
            "OpenBangla-Keyboard_2.0.0-fedora36.rpm",
            "2.0.0".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
            DateTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(pack.version(), "2.0.0");
        assert_eq!(
            *pack.distribution(),
            Some(Dist::Fedora(Some("36".to_owned())))
        );
        assert_eq!(*pack.ty(), Type::Rpm);

        let pack = Package::detect_package(
            "caprine_2.56.1_amd64.deb",
            "v2.56.1".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
            DateTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(pack.version(), "v2.56.1");
        assert_eq!(*pack.distribution(), None);
        assert_eq!(*pack.ty(), Type::Deb);
    }

    #[test]
    fn test_package_change_propagation() {
        let pack = Package::detect_package(
            "OpenBangla-Keyboard_2.0.0-ubuntu22.04.deb",
            "2.0.0".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
            DateTime::UNIX_EPOCH,
        )
        .unwrap();

        assert!(!pack.is_metadata_available());

        let pack2 = pack.clone();
        pack2.set_metadata(String::new());

        assert!(pack.is_metadata_available());
    }

    #[test]
    fn test_split_extension() {
        assert_eq!(
            split_extention("OpenBangla-Keyboard_2.0.0-ubuntu22.04.deb"),
            Some((Type::Deb, "OpenBangla-Keyboard_2.0.0-ubuntu22.04"))
        );
        assert_eq!(
            split_extention("OpenBangla-Keyboard_2.0.0-fedora36.rpm"),
            Some((Type::Rpm, "OpenBangla-Keyboard_2.0.0-fedora36"))
        );
        assert_eq!(split_extention("caprine_2.56.1_amd64.snap"), None);
        assert_eq!(split_extention("deb"), None);
    }

    #[test]
    fn test_split_test() {
        assert_eq!(split_at_numeric("ubuntu24.10"), Some("24.10"));
        assert_eq!(split_at_numeric("ubuntu"), None);
    }

    #[test]
    fn test_parse_version() {
        assert_eq!(parse_version("ubuntu22.10").unwrap(), "22.10".to_owned());
        assert_eq!(parse_version("fedora37").unwrap(), "37".to_owned());
    }
}
