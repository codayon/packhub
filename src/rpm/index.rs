use askama::Template;

use super::package::{Dependency, RPMPackage};

#[derive(Debug, Template)]
#[template(path = "primary.xml")]
struct Primary {
    packages: Vec<RPMPackage>,
}

#[derive(Debug, Template)]
#[template(path = "filelists.xml")]
struct FileLists {
    packages: Vec<RPMPackage>,
}

#[derive(Debug, Template)]
#[template(path = "other.xml")]
struct Other {
    packages: Vec<RPMPackage>,
}

pub fn get_primary_index(packages: Vec<RPMPackage>) -> String {
    let primary = Primary { packages };
    primary.render().unwrap()
}

pub fn get_filelists_index(packages: Vec<RPMPackage>) -> String {
    let list = FileLists { packages };
    list.render().unwrap()
}

pub fn get_other_index(packages: Vec<RPMPackage>) -> String {
    let list = Other { packages };
    list.render().unwrap()
}

#[cfg(test)]
mod tests {
    use std::fs::read;

    use chrono::DateTime;
    use insta::assert_snapshot;

    use crate::package::Package;

    use super::*;

    #[test]
    fn test_rpm_indices() {
        let package = Package::detect_package("OpenBangla-Keyboard_2.0.0-fedora38.rpm", "2.0.0".to_owned(), "https://github.com/OpenBangla/OpenBangla-Keyboard/releases/download/2.0.0/OpenBangla-Keyboard_2.0.0-fedora38.rpm".to_owned(), DateTime::UNIX_EPOCH).unwrap();
        let data = read("data/OpenBangla-Keyboard_2.0.0-fedora38.rpm").unwrap();
        package.set_data(data);
        let package = RPMPackage::from_package(&package).unwrap();
        let packages = vec![package];

        assert_snapshot!(get_primary_index(packages.clone()));

        assert_snapshot!(get_filelists_index(packages.clone()));

        assert_snapshot!(get_other_index(packages));
    }
}
