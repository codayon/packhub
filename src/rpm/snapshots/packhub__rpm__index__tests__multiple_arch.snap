---
source: src/rpm/index.rs
expression: get_primary_index(&packages)
---
<?xml version="1.0" encoding="UTF-8"?>
<metadata xmlns="http://linux.duke.edu/metadata/common" xmlns:rpm="http://linux.duke.edu/metadata/rpm" packages="7">

<package type="rpm">
  <name>fastfetch</name>
  <arch>aarch64</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">0318870bee582652669cbff9f41c21abc91eb9e98b12c8d0265a2332d4efb0cf</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <url>https://github.com/fastfetch-cli/fastfetch</url>
  
  <time file="0" build="1743815923"/>
  <size package="14349" installed="4453126" archive="4461824"/>
  <location href="package/2.40.3/fastfetch-linux-aarch64.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>fv-az888-37.dvsane4ksfzelhofvpftuz2bdb.xx.internal.cloudapp.net</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14349"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
      
      <rpm:entry name="fastfetch(aarch-64)" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="ld-linux-aarch64.so.1()(64bit)"/>
      
    
      
      <rpm:entry name="ld-linux-aarch64.so.1(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.33)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.34)(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

<package type="rpm">
  <name>fastfetch</name>
  <arch>x86_64</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">b86a9564343ca858c5a5a1c7942964917255882853ad844de69b2ef07e12e99d</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <url>https://github.com/fastfetch-cli/fastfetch</url>
  
  <time file="0" build="1743815950"/>
  <size package="14604" installed="29484355" archive="29493112"/>
  <location href="package/2.40.3/fastfetch-linux-amd64.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>fv-az772-168.zboo4qir1psufltgxvkvdwdv1g.dx.internal.cloudapp.net</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14604"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
      
      <rpm:entry name="fastfetch(x86-64)" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="libc.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.14)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.2.5)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.3)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.3.4)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.4)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.7)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.8)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.9)(64bit)"/>
      
    
      
      <rpm:entry name="libdl.so.2()(64bit)"/>
      
    
      
      <rpm:entry name="libdl.so.2(GLIBC_2.2.5)(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.2.5)(64bit)"/>
      
    
      
      <rpm:entry name="libpthread.so.0()(64bit)"/>
      
    
      
      <rpm:entry name="libpthread.so.0(GLIBC_2.2.5)(64bit)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

<package type="rpm">
  <name>fastfetch</name>
  <arch>armv6l</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">60f253a11ca3c8cfc367170ff3095298a009c974e346b1d0a76fbce150aae10e</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <url>https://github.com/fastfetch-cli/fastfetch</url>
  
  <time file="0" build="1743816853"/>
  <size package="14173" installed="16604476" archive="16613172"/>
  <location href="package/2.40.3/fastfetch-linux-armv6l.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>1f844b9c7a7b</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14173"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
      
      <rpm:entry name="fastfetch(armv6l-32)" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="libc.so.6"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.34)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.7)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.9)"/>
      
    
      
      <rpm:entry name="libm.so.6"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

<package type="rpm">
  <name>fastfetch</name>
  <arch>armv7l</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">6d3b0b5ea04abe7f5c92f0908f1819ac418ad76cef9b0162edc87863bdb895dc</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <url>https://github.com/fastfetch-cli/fastfetch</url>
  
  <time file="0" build="1743816905"/>
  <size package="14288" installed="22705849" archive="22714604"/>
  <location href="package/2.40.3/fastfetch-linux-armv7l.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>edfb8fea83b9</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14288"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="ld-linux-armhf.so.3"/>
      
    
      
      <rpm:entry name="ld-linux-armhf.so.3(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="libc.so.6"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.17)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.7)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.8)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.9)"/>
      
    
      
      <rpm:entry name="libdl.so.2"/>
      
    
      
      <rpm:entry name="libdl.so.2(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="libm.so.6"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="libpthread.so.0"/>
      
    
      
      <rpm:entry name="libpthread.so.0(GLIBC_2.4)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

<package type="rpm">
  <name>fastfetch</name>
  <arch>ppc64le</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">fd34074b0d87d03c7a0e8d606d0d6bac73b693e8db3b1d572ab86eb04ba3bcfc</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <time file="0" build="1743816850"/>
  <size package="14332" installed="28956387" archive="28965144"/>
  <location href="package/2.40.3/fastfetch-linux-ppc64le.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>d6e92c211299</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14332"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
      
      <rpm:entry name="fastfetch(ppc-64)" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="libc.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)(64bit)"/>
      
    
      
      <rpm:entry name="libdl.so.2()(64bit)"/>
      
    
      
      <rpm:entry name="libdl.so.2(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libpthread.so.0()(64bit)"/>
      
    
      
      <rpm:entry name="libpthread.so.0(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

<package type="rpm">
  <name>fastfetch</name>
  <arch>s390x</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">9578dc2b4d1e60e72f6540f00dc4b83af4198ce6f44079880bed3be9f653d47b</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <time file="0" build="1743816792"/>
  <size package="14572" installed="27865187" archive="27873944"/>
  <location href="package/2.40.3/fastfetch-linux-s390x.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>172a5f6f4874</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14572"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
      
      <rpm:entry name="fastfetch(s390-64)" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="/bin/sh"/>
      
    
      
      <rpm:entry name="libc.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.2)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.3)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.3.4)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.4)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.7)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.8)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.9)(64bit)"/>
      
    
      
      <rpm:entry name="libdl.so.2()(64bit)"/>
      
    
      
      <rpm:entry name="libdl.so.2(GLIBC_2.2)(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.2)(64bit)"/>
      
    
      
      <rpm:entry name="libpthread.so.0()(64bit)"/>
      
    
      
      <rpm:entry name="libpthread.so.0(GLIBC_2.2)(64bit)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

<package type="rpm">
  <name>fastfetch</name>
  <arch>riscv64</arch>
  <version epoch="0" ver="2.40.3" rel="1"/>
  <checksum type="sha256" pkgid="YES">3c65954b1de511fa09166f46145c443e7859bc86b6972ea891b554c9677b8fa4</checksum>
  
  <summary>Fast neofetch-like system information tool</summary>
  
  
  <description>DESCRIPTION
===========

This is an installer created using CPack (https://cmake.org). No additional installation instructions provided.</description>
  
  <packager></packager>
  
  <url>https://github.com/fastfetch-cli/fastfetch</url>
  
  <time file="0" build="1743816724"/>
  <size package="14373" installed="2802518" archive="2811216"/>
  <location href="package/2.40.3/fastfetch-linux-riscv64.rpm"/>
  <format>
    
    <rpm:license>MIT</rpm:license>
    
    
    <rpm:vendor>Humanity</rpm:vendor>
    
    
    <rpm:group>unknown</rpm:group>
    
    
    <rpm:buildhost>4f4f03450c46</rpm:buildhost>
    
    
    <rpm:sourcerpm>fastfetch-2.40.3-1.src.rpm</rpm:sourcerpm>
       
    <rpm:header-range start="4504" end="14373"/>
    <rpm:provides>
    
      
      <rpm:entry name="fastfetch" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
      
      <rpm:entry name="fastfetch(riscv-64)" flags="EQ" epoch="0" ver="2.40.3-1" rel="1"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="ld-linux-riscv64-lp64d.so.1()(64bit)"/>
      
    
      
      <rpm:entry name="ld-linux-riscv64-lp64d.so.1(GLIBC_2.27)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.27)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.33)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.34)(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.27)(64bit)"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

</metadata>
